packages/core/src/event/event.ts

```diff
@@ -0,0 +1,72 @@
+import * as t from 'io-ts';
+import { PathReporter } from 'io-ts/lib/PathReporter';
+import * as E from 'fp-ts/lib/Either';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { identity, flow } from 'fp-ts/lib/function';
+import { Event, EventWithPayload, EventWithoutPayload } from './event.interface';
+
+type Payload =
+  | t.TypeC<any>
+  | t.TupleC<any>
+  | t.ReadonlyArrayC<any>
+  | t.ArrayC<any>
+  | t.StringC
+  | t.BooleanC
+  | t.NumberC
+  | t.UndefinedC
+  ;
+
+export type EventSchema<T extends string> =
+  t.TypeC<{ type: t.LiteralC<T>; payload: t.UndefinedC }>
+
+export type EventSchemaWithPayload<T extends string, P extends Payload> =
+  t.TypeC<{ type: t.LiteralC<T>; payload: P }>
+
+type EventCreator<T extends string> =
+  () => EventWithoutPayload<T>
+
+type EventCreatorWithPayload<T extends string, P extends Payload> =
+  (payload: t.TypeOf<P>) => EventWithPayload<t.TypeOf<P>, T>
+
+type EventBuilderOutput<T extends string, P extends Payload | undefined> =
+  undefined extends P
+    ? EventSchema<T> & { create: EventCreator<T> }
+    : EventSchemaWithPayload<T, NonNullable<P>> & { create: EventCreatorWithPayload<T, NonNullable<P>> }
+
+/**
+ * Creates an Event codec for decoding and creating I/O events.
+ * @since v3.3.0
+ * @example
+ *
+ * const CreateUser = event('CREATE_USER')(t.type({
+ *   name: t.string,
+ *   age: t.number,
+ * }));
+ *
+ * CreateUser.create({
+ *   name: 'Bob',
+ *   age: 21,
+ * });
+ */
+export const event = <T extends string>(type: T) => <P extends Payload | undefined>(payload?: P): EventBuilderOutput<T, P> => {
+  const schema = t.type({
+    type: t.literal(type),
+    payload: payload ?? t.undefined,
+  }, 'EventSchema');
+
+  const create = (payload?: any): Event => {
```

Comment:
So, just for my own understanding, `create` would basically be a smart-constructor of the event type that avoids the `Either` in the result, and throws instead? That is, `decode`, without the `Either`?

Reply:
Basically yes, since it infers the payload type from schema. Right know I think that decoding the payload inside the constructor is a bit redundant and unnecessary. Basically it should act as a replacement for `encode` with the ability to just pass the payload object since the rest can be taken from the schema context. 🤔

Reply:
Which simplifies it to:
```typescript
  const create = (payload?: any): Event =>
    schema.encode({ type, payload });
```

Reply:
But again, maybe we can also overload the `encode` function and use it in two ways 🤔.
WDYT?

Reply:
I think `create` is sensible as a separate function, I like it how it is. I worry that overloading `encode` leads to misunderstanding for those that are used to `io-ts` codecs. But I believe there is value in documenting that `event` includes `encode` and `decode`.

Reply:
It could also be useful to document that `create` can throw, which, again, might be confusing to those used to `io-ts`.

Reply:
Ok then, `create` can be as it is but anyway I would like to remove this unnecessary decoding underneath and replace it with simple `encode`